### PR TITLE
Support vector tile source for additional viz types

### LINKED ISSUE #71: Enable `data` for any viz to come from a vector tile source
# Problem

Currently, each viz class `data` object only supports geojson format data from a URL or local variable geoJSON format object.  We should enable users to pass a data as a vector tile source.

# Implementation

I propose we maintain that the `data` object should be a URL or local geoJSON format object.  We can add another optional parameter to each parent viz class named `vector_source_layername`.  This will parameter specifies the name of the vector tile layer to create the visual from a layer in a vector tile.  

If the user passes the optional `vector_source_layername` parameter, the visualization will create a layer from a vector tile source instead of a geojson source.

### PR DESCRIPTION
**Pull request by akacarlyann**
Adds support for vector tile source for ```MapViz```, ```CircleViz```, ```GraduatedCircleViz``` and ```HeatmapViz``` via the ```VectorMixin``` class.  Adds support for allowing JSON/GeoJSON join-data to be passed as ```data``` argument as a filename, URL, or dictionary.  Closes #71.

### CODE REVIEW

**Comment by akacarlyann** [comment]
@ryanbaumann Still working on tests and docs for this PR, but it's pretty close :)

**Comment by akacarlyann** [comment]
@ryanbaumann Can you take a look and see if the tests I've included are sufficient?  Still need to update docs, but I should be able to get this done before the weekend :)

**Comment by ryanbaumann** [comment]
@akacarlyann looks like the external vector tile support for circle, graduatedcircle, and heatmap layers is all based on using the data-join technique.  One use case we want to also support is to style data based on properties already in the vector tile itself - is that possible with your new branch here?  

For example, if all we wanted to do was replace the `geojson` sources in this example https://nbviewer.jupyter.org/github/mapbox/mapboxgl-jupyter/blob/master/examples/notebooks/point-viz-types-example.ipynb with vector tile sources.  No data-join required from a local data frame, just sourcing the data from an external vector tile source.

**Comment by akacarlyann** [comment]
Sure, that was working in one of the intermediate steps. I'll make sure it's still working with the idea be that ```data``` can be passed in as ```[]``` or ```None```. I think the heatmap example still works with an empty list for data maybe...

**Comment by akacarlyann** [comment]
Hey @ryanbaumann , I've added a flag for disabling vector data join.  I'm also working on making sure join data is supported when passed as JSON/GeoJSON, a dictionary, or a filepath, so feel free to take a look.

File: docs/viz.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ Parameter | Description | Example
 data | GeoJSON Feature Collection or JSON Join-Data | 'points.geojson'
 vector_url | optional property to define vector data source (supported for basic MapViz, CircleViz, GraduatedCircleViz, HeatmapViz, ChoroplethViz, LinestringViz) | 'mapbox://mapbox.mapbox-terrain-v2'
 vector_layer_name | property to define target layer of vector source | 'contour'
-vector_join_property | property to aid in determining color for styling vector layer | 'ele'
+vector_join_property | property of features in vector tile data to use as link to joined json data | 'ele'
```
Comment (ryanbaumann) [inline docs/viz.md]: `Name of the property in the reference vector tile source to join to the `JSON` data passed to this function

**Comment by ryanbaumann** [comment]
Going to try testing this now on various viz types @akacarlyann

**Comment by akacarlyann** [comment]
@ryanbaumann Have you had a chance to take a look at this? Hoping to spend some time on the project later this week :)

**Comment by ryanbaumann** [comment]
Awesome @akacarlyann.  Unfortunately no, I haven't had much time to check this out.  Thank you for the reminder - I'll do it ASAP.

**Comment by ryanbaumann** [comment]
Largely this looks good @akacarlyann and will be super powerful for adding vector data sources.

One question - could eliminate the need for the `disable_data_join=True` flag when using a vector tile source?  It should be the default behavior that most users will want if they are not making a Choropleth map.

**Comment by akacarlyann** [comment]
@ryanbaumann Can you explain a little more of what you mean? Under the current vector visualization paradigm, the user has to pass a json `data` object and the vector tile URL and layer name. But if a user wants to style the layer only using the data within the vector tile source, they must still provide json data for the layer filter applied to the vector data. `disable_data_join=True` just prevents the empty `data` param from filtering out all the features in the vector source.

I can refactor this but it will take a little more logic to smartly determine when `data` should be ignored. Alternately, join data could be added in a separate argument?

**Comment by ryanbaumann** [comment]
> But if a user wants to style the layer only using the data within the vector tile source, they must still provide json data for the layer filter applied to the vector data. disable_data_join=True just prevents the empty data param from filtering out all the features in the vector source.

Understood @akacarlyann.  I think this is a reasonable assumption to make then, as long as we document it well, which you have.

Checking out the branch for merge now with the latest changes merged into master.

**Comment by akacarlyann** [comment]
@ryanbaumann looks like one check is failing now due to updates on master in the last few weeks. I'll look into fixing that, but otherwise, how is this PR looking?

**Comment by ryanbaumann** [comment]
@akacarlyann looking good with a master update - thank you!  Ill approve once passing.

**Comment by akacarlyann** [comment]
@ryanbaumann ready to go!

**Comment by ryanbaumann** [comment]
Woho!  Thanks for bearing with me on the time to merge - this opens up a huge number of use cases visualizing massive pre-tiled data sources in Jupyter and Python.  Thanks @akacarlyann!

**Pull request merged**